proto_proxy.c

```diff
@@ -0,0 +1,2310 @@
+    if (ss->phc->ctx == NULL) {
+        // TODO: if NULL, pass in pool_size as ctx?
+        // works because the % bit will return an id we can index here.
+        // FIXME: temporary? maybe?
+        // if no context, what we got back was a hash which we need to modulus
+        // against the pool, since the func has no info about the pool.
+        rq->be = ss->pool[lookup % ss->pool_size].be;
+    } else {
+        // else we have a direct id into our pool.
+        // the lua modules should "think" in 1 based indexes, so we need to
+        // subtract one here.
+        // TODO: bother validating the range?
+        rq->be = ss->pool[lookup-1].be;
+    }
+
+    // now yield request, hash selector up.
+    return lua_yield(L, 2);
+}
+
+// mcp.attach(mcp.HOOK_NAME, function|userdata)
+// fill hook structure: if lua function, use luaL_ref() to store the func
+// if it a userdata of the proper type, set its C function + data pointers
+// for direct callback.
+static int mcplib_attach(lua_State *L) {
+    // Pull the original worker thread out of the shared mcplib upvalue.
+    LIBEVENT_THREAD *t = lua_touserdata(L, lua_upvalueindex(MCP_THREAD_UPVALUE));
+
+    int hook = luaL_checkinteger(L, -2);
+    if (lua_isuserdata(L, -1)) {
+        // TODO: not super sure how to create the API here.
+        // it needs/should identify to a specific userdata type so we can
+        // generically recover the function and data pointer.
+    } else if (lua_isfunction(L, -1)) {
+        t->proxy_hook = hook;
+
+        if (t->proxy_attach_ref) {
+            luaL_unref(L, LUA_REGISTRYINDEX, t->proxy_attach_ref);
+        }
+
+        // pops the function from the stack and leaves us a ref. for later.
+        t->proxy_attach_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+    } else {
+        // TODO: throw error.
+    }
+
+    return 0;
+}
+
+// TODO: temporary defines.
+// - any reason to not use these?
+#define CMD_FIELDS \
+    X(CMD_MG) \
+    X(CMD_MS) \
+    X(CMD_MD) \
+    X(CMD_MN) \
+    X(CMD_MA) \
+    X(CMD_ME) \
+    X(CMD_GET) \
+    X(CMD_GAT) \
+    X(CMD_SET) \
+    X(CMD_ADD) \
+    X(CMD_CAS) \
+    X(CMD_LRU) \
+    X(CMD_GETS) \
+    X(CMD_GATS) \
+    X(CMD_INCR) \
+    X(CMD_DECR) \
+    X(CMD_QUIT) \
+    X(CMD_STATS) \
+    X(CMD_SLABS) \
+    X(CMD_TOUCH) \
+    X(CMD_WATCH) \
+    X(CMD_APPEND) \
+    X(CMD_DELETE) \
+    X(CMD_REPLACE) \
+    X(CMD_PREPEND) \
+    X(CMD_VERSION) \
+    X(CMD_SHUTDOWN) \
+    X(CMD_EXTSTORE) \
+    X(CMD_FLUSH_ALL) \
+    X(CMD_VERBOSITY) \
+    X(CMD_LRU_CRAWLER) \
+    X(CMD_REFRESH_CERTS) \
+    X(CMD_CACHE_MEMLIMIT)
+
+#define X(name) name,
+enum proxy_defines {
+    P_OK = 0,
+    CMD_ANY,
+    CMD_FIELDS
+};
+#undef X
+
+static void proxy_register_defines(lua_State *L) {
+#define X(x) \
+    lua_pushinteger(L, x); \
+    lua_setfield(L, -2, #x);
+
+    X(P_OK);
+    X(CMD_ANY);
+    CMD_FIELDS
+#undef X
+}
+
+/*** REQUEST PARSER AND OBJECT ***/
+
+// FIXME: kill prints.
```

Comment:
It may be possible to replace the prints with a macro instead (e.g. and include `__FILE__` and `__LINE__`), and make that macro a no-op if the proxy isn't compiled with a flag such as `--enable-proxy-debug-log` - e.g. just `ERROR\n` is ambiguous on its own when locally developing or receiving a bug report, but `proto_proxy.c:1234 ERROR` is more specific

Separating errors/warnings/info/debug levels may also be useful when debugging.

Reply:
That's what `P_DEBUG` is. I didn't add the flags yet and need to refactor it to move __func__ inside P_DEBUG to shorten the commands. I've been converting prints to P_DEBUG where it makes sense.